### PR TITLE
Add script to re-pxe VM clusters.

### PR DESCRIPTION
**Pull request by http-418**

### CODE REVIEW

**Comment by cbaenziger** [comment]
Hi Andrew, I hit the following issue following the steps in `repxe_host.rb`:
````
vagrant@bcpc-bootstrap:~/chef-bcpc$ /opt/chef/embedded/bin/bundle install --deployment
The --deployment flag requires a Gemfile.lock. Please make sure you have checked your Gemfile.lock into version control before deploying.
````

**Comment by cbaenziger** [comment]
Next, I hit a MITM proxy issue; I modified the gemsource from https to http and source'd `proxy_setup.sh` which then reported:
````
vagrant@bcpc-bootstrap:~/chef-bcpc$ /opt/chef/embedded/bin/bundle install 
Fetching gem metadata from http://rubygems.org/..............
Resolving dependencies...

Gem::InstallError: amq-protocol requires Ruby version >= 2.0.
An error occurred while installing amq-protocol (2.0.1), and Bundler cannot continue.
Make sure that `gem install amq-protocol -v '2.0.1'` succeeds before bundling.
````
How have you been installing Ruby 2.0?

**Comment by cbaenziger** [comment]
Hrm, trying to hack a `Gemfile` that would work with chef-provisioning and Chef 11 (aka Ruby 1.9) seem neigh impossible. I grabbed a ChefDK (0.10.0) to grab a Ruby 2.x and am now hitting the following:
````
vagrant@bcpc-bootstrap:~/chef-bcpc$ /opt/chefdk/embedded/bin/bundle install
[...]
Bundle complete! 7 Gemfile dependencies, 49 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
vagrant@bcpc-bootstrap:~/chef-bcpc$ /opt/chefdk/embedded/bin/bundle exec ./repxe_host.rb 10.0.100.13
Could not find erubis-2.7.0 in any of the sources
Run `bundle install` to install missing gems.
````
I also hit an issue with the `pry` gem. I modified the Gemfile with:
````
+gem 'erubis', '~> 2.7.0'
+gem 'pry'
````

I set my PATH to `export PATH=/opt/chefdk/embedded/bin:$PATH` and ran:
`vagrant@bcpc-bootstrap:~/chef-bcpc$ /opt/chefdk/embedded/bin/bundle exec ./repxe_host.rb bcpc-vm3` but hit:
````
Un-enrolled bcpc-vm3 from cobbler
Found 08:00:27:29:35:07 for bcpc-vm3 (10.0.100.13)
Enrolled bcpc-vm3 in cobbler
Deleted SSH fingerprints and Chef objects for bcpc-vm3
Please reboot bcpc-vm3, then hit enter

./repxe_host.rb:109:in `cobbler_root_password': Could not retrieve cobbler password!\n\nWARN: Ohai::Config[:disabled_plugins] is set. Ohai::Config[:disabled_plugins] is deprecated and will be removed in future releases of ohai. Use ohai.disabled_plugins in your configuration file to configure :disabled_plugins for ohai. (RuntimeError)
ERROR: SSL Validation failure connecting to host: 10.0.100.3 - SSL_connect returned=1 errno=0 state=error: certificate verify failed
ERROR: Could not establish a secure connection to the server.
Use `knife ssl check` to troubleshoot your SSL configuration.
If your Chef Server uses a self-signed certificate, you can use
`knife ssl fetch` to make knife trust the server's certificates.

Original Exception: OpenSSL::SSL::SSLError: SSL_connect returned=1 errno=0 state=error: certificate verify failed
        from ./repxe_host.rb:253:in `wait_for_host'
        from ./repxe_host.rb:314:in `<main>'
````

**Comment by cbaenziger** [comment]
Adding the following to `.chef/knife.rb` seems to have helped `verify_api_cert false`

This seems like a bad failure mode. (I.e. if we need to re-PXE a machine which is down and not reachable we won't be able to get an ARP entry.)
````
Ping to bcpc-vm3 (10.0.100.13) failed, checking ARP anyway.
./repxe_host.rb:149:in `corrected_mac': Could not find ARP entry for bcpc-vm3 (10.0.100.13)! (RuntimeError)
        from ./repxe_host.rb:84:in `cobbler_enroll'
        from ./repxe_host.rb:310:in `<main>'
````

We should trust `cluster.txt` to have the right MAC; unless I'm missing the point?

**Comment by http-418** [comment]
Clay, the newest commit should fix your issues with Ruby 1.9 and Bundler.

**Comment by http-418** [comment]
Re: ARP, for testing a VM cluster, I found it convenient to add a static entry:
`sudo arp -s 10.0.100.14 08:00:27:80:c0:8f`

File: repxe_host.rb [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+require 'chef/provisioning/transport/ssh'
+require 'mixlib/shellout'
+require 'pry'
+require 'timeout'
+
+def cluster_assign_roles(environment,type,entry=nil)
+  if(type.to_s.downcase != 'basic' &&
+     type.to_s.downcase != 'hadoop')
+    raise "#{type} is not one of 'basic', 'hadoop' !"
```
Comment (cbaenziger) [inline repxe_host.rb]: Should support `kafka` too...

File: repxe_host.rb [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+  end
+
+  # We use system() instead of Mixlib::ShellOut specifically so that the
+  # child process re-uses our STDOUT/STDERR.
```
Comment (cbaenziger) [inline repxe_host.rb]: I thought one could setup pipes for I/O redirection to allow this to still be tested?

**Comment by cbaenziger** [comment]
I'm hitting the following issue:
{code}
Please reboot cluster1-r1n1, then hit enter

./repxe_host.rb:115:in `cobbler_root_password': Could not retrieve cobbler password!\nERROR: SSL Validation failure connecting to host: 1.2.3.4 - SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (RuntimeError)
\nERROR: Could not establish a secure connection to the server.
Use `knife ssl check` to troubleshoot your SSL configuration.
If your Chef Server uses a self-signed certificate, you can use
`knife ssl fetch` to make knife trust the server's certificates.

Original Exception: OpenSSL::SSL::SSLError: SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed
        from ./repxe_host.rb:259:in `wait_for_host'
        from ./repxe_host.rb:320:in `<main>'
{code}

File: repxe_host.rb [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    exit(-1)
+  end
+
+  cobbler_unenroll(vm_entry)
+  cobbler_enroll(vm_entry)
+  cobbler_sync
+  delete_node_data(vm_entry)
+  restart_host(vm_entry)
+  wait_for_host(vm_entry)
+  cluster_assign_roles('Test-Laptop', :basic, vm_entry)
```
Comment (cbaenziger) [inline repxe_host.rb]: This will need to pick the environment the bootstrap's in; not `Test-Laptop`

File: repxe_host.rb (hunk 2 above)
Comment (http-418) [inline repxe_host.rb]: Not with mixlib::shellout.   There's a popen3 family of calls to customize I/O redirection, and I guess that's hypothetically more testable, but I'm happy with the default behavior, and I do not expect to write tests for this code.

File: repxe_host.rb (hunk 1 above)
Comment (http-418) [inline repxe_host.rb]: I added Kafka.

I shouldn't call the shell script with invalid parameters.

File: .gitignore [hunk 4]
```diff
@@ -20,4 +22,3 @@ cookbooks/*/
 !cookbooks/bach_krb5
 cookbooks/*/coverage
 cookbooks/*/vendor
-*.lock
```
Comment (cbaenziger) [inline .gitignore]: Let's leave the `*.lock` ignore in place?

**Comment by http-418** [comment]
Esther has discovered that removing the lockfile actually breaks the instructions, so I will need to fix that.

**Comment by ekund** [comment]
@http-418 It would be good if the pause that asks for the reboot explicitly mentioned that it's time to reboot the machine in pxe-mode as a reminder.

**Comment by ekund** [comment]
I've found a few issues.  1, some of the order of the steps was not exactly as specified, though I'm not sure if it helped.  Also, I believe this line:
`cluster_assign_roles(chef_env, :hadoop)`
needs to be removed.  It attempts to run CAR on every node in the cluster, which then fails, and I believe is unnecessary.

**Pull request closed without merging**